operator/apis/loki/v1/lokistack_types.go

```diff
@@ -345,6 +345,58 @@ type ClusterProxy struct {
 	NoProxy string `json:"noProxy,omitempty"`
 }
 
+// HashRingType defines the type of hash ring which can be used with the Loki cluster.
+//
+// +kubebuilder:validation:Enum=memberlist
+type HashRingType string
+
+const (
+	// HashRingMemberList when using memberlist for the distributed hash ring.
+	HashRingMemberList HashRingType = "memberlist"
+)
+
+// MemberListNetworkType defines the type of pod network to use for advertising IPs to the ring.
+type MemberListNetworkType string
+
+const (
+	// MemberListPrivateNetwork when using the private pod networks (RFC 1918 and RFC 6598)
+	MemberListPrivateNetwork MemberListNetworkType = "private"
+	// MemberListPublicNetwork when using the public pod networks, i.e will use status.podIP
+	MemberListPublicNetwork MemberListNetworkType = "public"
+)
+
+// MemberListSpec defines the configuration for the memberlist based hash ring.
+type MemberListSpec struct {
+	// BindNetworkType defines the type of network to use to advertise IPs to the ring.
+	// Defaults to the private network available to the current pod. Alternatively the
+	// pod public network can be used in case private networks (RFC 1918 and RFC 6598)
+	// are not available.
+	//
+	// +required
+	// +kubebuilder:validation:Required
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:select:private","urn:alm:descriptor:com.tectonic.ui:select:public"},displayName="Hash Ring Type"
+	// +kubebuilder:default:=private
+	BindNetworkType MemberListNetworkType `json:"bindNetworkType"`
```

Comment:
> Two questions come up because of this:
> 
>     * Do we actually have to change the bind? The memberlist complains of not finding an "advertise address", looking at the [Loki docs](https://grafana.com/docs/loki/latest/configuration/) the `instance_addr` configuration field seems like a better choice for specifically setting the address advertised in the ring. (Have not looked at the actual implementation yet though, so keep a grain of salt ready.)

I have to admit that `instance_addr` is new to me and in general the `common.ring` section is not obvious to me on how it interacts with the `memberlist_config`. Later is/was the no.1 section for the gossip ring configured once for all ring-related component sets: 1) distributor, ingester, 2) querier, ingester,  3) ruler and 4) compactor (last not used in the Loki Operator as per compactor single replica for all tenants applied). 

FWIW I need to dig into the Loki code base in how `instance_addr` and `bind_addr` interact with each other. Another disturbing fact about the `instance_addr` setting is the following comment on the Loki docs:

```
# If "instance_addr" under the common ring section is configured, this common "instance_addr"
# is only applied to the frontend, but not for ring related components (ex: distributor, ruler, etc).
```

> 
>     * If we want to keep setting the bind address: Does it make sense to make the setting "more explicit", for example by having `any` and `podIP` as possible settings instead of the current `private`/`public`?

I believe your suggestion is easier to comprehend and will consider switching to enum options `any` and `podIP` after investigating the interaction of `instance_addr` and `bind_addr`.

Reply:
Leaving this as a note here, none conclusions yet, but it looks like `bind_addr` is the first fallback if we don't provide `instance_addr` in the ring configuration for `memberlist`: https://github.com/grafana/dskit/blob/main/kv/memberlist/tcp_transport.go#L347-L397